modules/core/src/smithy4s/codecs/FieldFilter.scala

```diff
@@ -0,0 +1,176 @@
+  }
+
+  private trait SkipNonRequired extends FieldFilter {
+
+    final def compile[S, A](
+        field: Field[S, A]
+    ): FieldFilter.Predicate[A] = {
+      if (field.isRequired) Function.const(true)
+      else compileOptional(field)
+
+    }
+
+    def compileOptional[S, A](
+        field: Field[S, A]
+    ): FieldFilter.Predicate[A]
+
+  }
+
+  case object EncodeAll extends FieldFilter {
+    def compile[S, A](field: Field[S, A]): Predicate[A] = Function.const(true)
+  }
+
+  private def asNonEmptyCollectionPredicate[F[_], A](
+      schema: Schema[A]
+  ): Option[A => Boolean] = {
+    import Schema._
+    schema match {
+      case c: CollectionSchema[f, a] =>
+        Some((collectionA: f[a]) => !c.tag.isEmpty(collectionA))
+      case b: BijectionSchema[inner, a] =>
+        asNonEmptyCollectionPredicate[F, inner](b.underlying)
+          .map(_.compose(b.bijection.from))
+      case r: RefinementSchema[inner, a] =>
+        asNonEmptyCollectionPredicate[F, inner](r.underlying)
+          .map(_.compose(r.refinement.from))
+      case o: OptionSchema[inner] =>
+        asNonEmptyCollectionPredicate(o.underlying)
+          .map(predicateInner =>
+            collectionA => collectionA.exists(predicateInner)
+          )
+      case _: MapSchema[k, v] =>
+        Some(collectionA => collectionA.nonEmpty)
+      case LazySchema(suspend) =>
+        // it is safe to call .value here because we don't recurse into structs/unions schemas,
+        // so we never see the same schema twice in this visitor.
+        asNonEmptyCollectionPredicate(suspend.value)
+      case _: EnumerationSchema[_] => None
+      case _: StructSchema[_]      => None
+      case _: UnionSchema[_]       => None
+      case _: PrimitiveSchema[_]   => None
+    }
+  }
+
+  private case object skipIfEmptyOptionalCollection
+      extends FieldFilter.SkipNonRequired {
+
+    def compileOptional[S, A](field: Field[S, A]): Predicate[A] = {
+      asNonEmptyCollectionPredicate(field.schema) match {
+        case None             => Function.const(true)
+        case Some(isNonEmpty) => isNonEmpty
+      }
+    }
+  }
+
+  val SkipIfEmptyOptionalCollection: FieldFilter =
+    skipIfEmptyOptionalCollection
+
+  case object SkipIfEmptyCollection extends FieldFilter {
+
+    def compile[S, A](field: Field[S, A]): Predicate[A] = {
+      asNonEmptyCollectionPredicate(field.schema) match {
+        case None             => Function.const(true)
+        case Some(isNonEmpty) => isNonEmpty
+      }
+    }
+  }
+
+  private case object skipDefaultOptionValues
+      extends FieldFilter.SkipNonRequired {
+    def compileOptional[S, A](field: Field[S, A]): Predicate[A] = {
+      // do not use field.hasDefaultValue as it returns always true for options
+      if (field.schema.getDefault.isDefined) { a =>
+        !field.isDefaultValue(a)
+      } else {
+        Function.const(true)
+      }
+    }
+  }
+
+  val SkipDefaultOptionValues: FieldFilter = skipDefaultOptionValues
```

Comment:
Rename to `SkipDefaultValues`

Reply:
Also why make a private case object and a value ?

Reply:
https://github.com/disneystreaming/smithy4s/pull/1652#discussion_r1967425666

Reply:
Sorry, I forgot about this one due to the other comment about private case objects. Applied.